src/buildstream/source.py

```diff
@@ -308,6 +308,15 @@ def mark_download_url(self, url: str) -> None:
 
         Args:
            url: The url used to download.
+
+        .. note::
+
+           While this must be called in a SourceFetcher initializer for the URL which
```

Comment:
Looks like this isn't consistently done. In buildstream-plugins, the `cargo` plugin does it but the `git` plugin doesn't. In bst-plugins-experimental, all plugins using source fetchers (`git_tag`, `git_repo` and `pypi`) do call it in the initializer.

I noticed however that the `git` plugin calls it from get_source_fetchers() instead, which is weird but probably gives the same end result?

Reply:
The `git` plugin technically *should* call `Source.mark_download_url()` *"for any URL that is specified in the source configuration"*, i.e. not necessarily for URLs which it will discover later by virtue of having the git module present and discovering submodules.

Technically without calling this, BuildStream core is unable to validate aliases which are used in the URL configurations in the yaml (although this *reason* need not be communicated as the sole purpose in the contract).

Reply:
I might have gotten things mixed up, I'll try to have a deeper look this evening.